agent/src/control_interface/authorizer.rs

```diff
@@ -0,0 +1,340 @@
+
+        Self {
+            allow_write_state_rule: allow_rules.write,
+            deny_write_state_rule: deny_rules.write,
+            allow_read_state_rule: allow_rules.read,
+            deny_read_state_rule: deny_rules.read,
+            allow_read_write_state_rule: allow_rules.read_write,
+            deny_read_write_state_rule: deny_rules.read_write,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+struct Rule<P: PathPattern> {
+    patterns: Vec<P>,
+}
+
+impl<P: PathPattern> PathPattern for Rule<P> {
+    fn matches(&self, path: &Path) -> bool {
+        self.patterns.iter().any(|p| p.matches(path))
+    }
+}
+
+// impl Rule {
+//     pub fn applies(&self, filter_mask: String, target_value: String) -> bool {
+//         self.filter_mask_applies(filter_mask) && self.targe_value_applies(target_value)
+//     }
+
+//     fn filter_mask_applies(&self, filter_mask: String) -> bool {
+//         self.filter_masks.is_empty() || self.filter_masks.iter().any(|x| x.matches(&filter_mask))
+//     }
+
+//     fn targe_value_applies(&self, target_value: String) -> bool {
+//         self.target_values.is_empty() || self.target_values.contains(&target_value)
+//     }
+// }
+
+#[derive(Clone, Debug)]
+struct Path {
+    sections: Vec<String>,
+}
+
+impl From<&str> for Path {
+    fn from(value: &str) -> Self {
+        Self {
+            sections: if value.is_empty() {
+                Vec::new()
+            } else {
+                value.split(PATH_SEPARATOR).map(Into::into).collect()
+            },
+        }
+    }
+}
+
+trait PathPattern {
+    fn matches(&self, other: &Path) -> bool;
+}
+
+impl<T: PathPattern> PathPattern for Vec<T> {
+    fn matches(&self, path: &Path) -> bool {
+        self.iter().any(|r| r.matches(path))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+struct AllowPathPattern {
+    sections: Vec<PathPatterSection>,
+}
+
+impl From<&str> for AllowPathPattern {
+    fn from(value: &str) -> Self {
+        Self {
+            sections: value.split(PATH_SEPARATOR).map(Into::into).collect(),
+        }
+    }
+}
+
+impl PathPattern for AllowPathPattern {
+    fn matches(&self, other: &Path) -> bool {
+        if self.sections.len() > other.sections.len() {
+            return false;
+        }
+        self.sections
+            .iter()
+            .zip(other.sections.iter())
+            .all(|(a, b)| a.matches(b))
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+struct DenyPathPattern {
+    sections: Vec<PathPatterSection>,
+}
+
+impl From<&str> for DenyPathPattern {
+    fn from(value: &str) -> Self {
+        Self {
+            sections: value.split(PATH_SEPARATOR).map(Into::into).collect(),
+        }
+    }
+}
+
+impl PathPattern for DenyPathPattern {
+    fn matches(&self, other: &Path) -> bool {
+        !self
+            .sections
+            .iter()
+            .zip(other.sections.iter())
+            .any(|(a, b)| !a.matches(b))
```

Comment:
A lot of "!" (not s) here even in the final evaluation above. Like discussed, this could be changed to a "all()) with removing the "!".

Reply:
I have rewritten this part, to provide logs explaining why a request has been allowed/denied.